File: src/types/security.ts
export type SecurityEventType =
  | 'authentication'
  | 'authorization'
  | 'data_access'
  | 'system_change'
  | 'api_access'
  | 'mfa'
  | 'user_management'
  | 'security_control'
  | 'compliance'
  | 'anomaly';

export type SecurityEventSeverity = 'low' | 'medium' | 'high' | 'critical';

export interface SecurityEvent {
  id: string;
  type: SecurityEventType;
  subtype?: string;
  severity: SecurityEventSeverity;
  timestamp: string;
  userId?: string;
  sessionId?: string;
  source: string;
  details: string;
  metadata?: Record<string, any>;
  relatedEvents?: string[];
  ipAddress?: string;
  userAgent?: string;
  status: 'detected' | 'investigating' | 'resolved' | 'false_positive';
  resolution?: string;
}

export interface SecurityAlert {
  id: string;
  eventId: string;
  timestamp: string;
  severity: SecurityEventSeverity;
  status: 'new' | 'acknowledged' | 'investigating' | 'resolved';
  assignedTo?: string;
  description: string;
  actions: SecurityAlertAction[];
  metadata?: Record<string, any>;
}

export interface SecurityAlertAction {
  id: string;
  alertId: string;
  type: 'notification' | 'remediation' | 'investigation';
  status: 'pending' | 'in_progress' | 'completed' | 'failed';
  timestamp: string;
  details: string;
  performedBy?: string;
  result?: string;
}

export interface SecurityMetrics {
  eventCounts: Record<SecurityEventType, number>;
  severityCounts: Record<SecurityEventSeverity, number>;
  timeBasedMetrics: {
    hourly: number[];
    daily: number[];
    weekly: number[];
  };
  topSources: Array<{ source: string; count: number }>;
  responseMetrics: {
    averageResponseTime: number;
    resolutionRate: number;
    falsePositiveRate: number;
  };
} 

File: src/types/security.py
from dataclasses import dataclass
from typing import Literal, Optional, Any, Dict, List, Union


SecurityEventType = Literal[
    'authentication',
    'authorization',
    'data_access',
    'system_change',
    'api_access',
    'mfa',
    'user_management',
    'security_control',
    'compliance',
    'anomaly'
]

SecurityEventSeverity = Literal['low', 'medium', 'high', 'critical']


@dataclass
class SecurityEvent:
    id: str
    type: SecurityEventType
    subtype: Optional[str] = None
    severity: SecurityEventSeverity
    timestamp: str
    userId: Optional[str] = None
    sessionId: Optional[str] = None
    source: str
    details: str
    metadata: Optional[Dict[str, Any]] = None
    relatedEvents: Optional[List[str]] = None
    ipAddress: Optional[str] = None
    userAgent: Optional[str] = None
    status: Literal['detected', 'investigating', 'resolved', 'false_positive']
    resolution: Optional[str] = None


@dataclass
class SecurityAlert:
    id: str
    eventId: str
    timestamp: str
    severity: SecurityEventSeverity
    status: Literal['new', 'acknowledged', 'investigating', 'resolved']
    assignedTo: Optional[str] = None
    description: str
    actions: List['SecurityAlertAction']
    metadata: Optional[Dict[str, Any]] = None


@dataclass
class SecurityAlertAction:
    id: str
    alertId: str
    type: Literal['notification', 'remediation', 'investigation']
    status: Literal['pending', 'in_progress', 'completed', 'failed']
    timestamp: str
    details: str
    performedBy: Optional[str] = None
    result: Optional[str] = None


@dataclass
class SecurityMetrics:
    eventCounts: Dict[SecurityEventType, int]
    severityCounts: Dict[SecurityEventSeverity, int]
    timeBasedMetrics: Dict[str, List[int]]
    topSources: List[Dict[str, Union[str, int]]]
    responseMetrics: Dict[str, Union[float, int]]